File: src/previous/day101/CompareBinaryTrees.ts
export default function compare(
  a: BinaryNode<number> | null,
  b: BinaryNode<number> | null
): boolean {
  if (!a && !b) return true
  if (!a || !b) return false
  if (a.value !== b.value) return false

  return compare(a.left, b.left) && compare(a.right, b.right)
}

File: go/day01/CompareBinaryTrees/CompareBinaryTrees.go
package compare_binary_tree

import data_structs "github.com/pakkermandev/kata-machine-learning/DataStructs"

func CompareBinaryTree[T comparable](a, b *data_structs.BinaryTree[T]) bool {
	var recurse func(a, b *data_structs.BinaryTreeNode[T]) bool
	recurse = func(a, b *data_structs.BinaryTreeNode[T]) bool {
		if a == nil && b == nil {
			return true
		}
		if a == nil || b == nil {
			return false
		}
		if a.Value != b.Value {
			return false
		}

		return recurse(a.Left, b.Left) && recurse(a.Right, b.Right)
	}

	return recurse(a.Root, b.Root)
}
